src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu

```diff
@@ -67,65 +70,139 @@ namespace cuda
         throw std::runtime_error("not implemented"); //TODO
     }
 
-    void RotateVisibilities(
-        Integration& integration,
-        MetaDataCudaHost& metadata,
-        const MVDirection& direction)
+    __device__ __forceinline__ cuDoubleComplex cuCexp(cuDoubleComplex z)
     {
-        //using namespace std::literals::complex_literals;
-        auto& data = integration.data;
-        auto& uvw = integration.uvw;
-        auto parameters = integration.parameters;
+        // see https://forums.developer.nvidia.com/t/complex-number-exponential-function/24696/2
+        cuDoubleComplex res;
+        sincos(z.y, &res.y, &res.x);
+        double t = exp(z.x);
+        res.x *= t;
+        res.y *= t;
+        return res;
+    }
 
-        if(metadata.init)
-        {
-            //metadata['nbaseline']=metadata['stations']*(metadata['stations']-1)/2
-            
-            metadata.SetDD(direction);
-            metadata.SetWv();
-            // Zero a vector for averaging in time and freq
-            metadata.avg_data = Eigen::MatrixXcd(integration.baselines, metadata.GetConstants().num_pols);
-            metadata.init = false;
-        }
-        metadata.CalcUVW(uvw);
+    __global__ void g_RotateVisibilities(
+        //Integration integration,
+        Constants constants,
+        Eigen::Matrix3d dd,
+        double2 direction,
+        double3* uvw, int uvwLength,
+        double3* oldUVW, int oldUVWLegth,
+        cuDoubleComplex* pavg_data, int avg_dataRows, int avg_dataCols)
+    {
+        using VectorXcucd = Eigen::Matrix<cuDoubleComplex, Eigen::Dynamic, 1>;
+        using MatrixXcucd = Eigen::Matrix<cuDoubleComplex, Eigen::Dynamic, Eigen::Dynamic>;
+        using Tensor3Xcucd = Eigen::Matrix<VectorXcucd, Eigen::Dynamic, Eigen::Dynamic>;
```

Comment:
Replacing this with `Eigen::Matrix<Eigen::Map<VectorXcucd>, Eigen::Dynamic, Eigen::Dynamic>` would almost work but cpu code still needs to allocate memory in a single contiguous block.

In https://github.com/ICRAR/leap-accelerate/pull/21 I refactor MetaData to instead use Eigen::Tensor. It's still possible to access a 2D tensor with the matrix methods using Eigen::Map<Eigen::MatrixXd> so I'm convinced it's the cleanest way to handle dynamic memory allocation.